homeassistant/components/stream/worker.py

```diff
@@ -202,6 +201,48 @@ def close(self) -> None:
         self._memory_file.close()
 
 
+class PeekIterator(Iterator):
+    """An Iterator that may allow multiple passes.
+
+    This may be consumed like a normal Iterator, however also supports a
+    peek() and rewind() methods that buffer consumed items from the iterator.
+    """
+
+    def __init__(self, iterator: Iterator[av.Packet]) -> None:
+        """Initialize PeekIterator."""
+        self._iterator = iterator
+        self._cursor: av.Packet = None
+        self._buffer: deque[av.Packet] = deque()
+
+    def __iter__(self) -> Iterator:
+        """Return an iterator."""
+        return self
+
+    def __next__(self) -> av.Packet:
```

Comment:
We only use `peek()` in the beginning, but after the beginning we don't need this functionality anymore. Although performance isn't a huge issue here it seems kind of wasteful to wrap the original iterator and check `self._buffer` after we're done using it. To avoid this, perhaps we can add a method which we can call after the peeking is done that removes the peeking functionality by replacing `self.__next__` with `self._iterator.__next__`

Reply:
I wasn't sure i could completely omit a __next__ given its an abstract method, though I tried to get as close as I could to what you're saying. Let me know if you have a better suggestion for how to do this.

Reply:
I was originally thinking of monkey patching `__next__`, but I guess that's not kosher and mypy complained. I then thought about using a method to return a new `itertools.chain()` iterator, similar to the prior behavior. But I think the way you've done it here (monkey patching automatically after the buffer has been consumed) is good. There's still the overhead of an extra wrapper in __next__, but it seems hard to avoid that while keeping the iterator clean (`itertools.chain()` will have that too).